six/modules/c++/cphd/source/SupportBlock.cpp

```diff
@@ -0,0 +1,156 @@
+#include <cphd/ByteSwap.h>
+#include <cphd/SupportBlock.h>
+#include <six/Init.h>
+
+namespace cphd
+{
+const std::string SupportBlock::ALL = six::Init::undefined<std::string>();
+
+
+SupportBlock::SupportBlock(const std::string& pathname,
+                           const cphd::Data& data,
+                           sys::Off_T startSupport,
+                           sys::Off_T sizeSupport) :
+    mInStream(new io::FileInputStream(pathname)),
+    mData(data),
+    mSupportOffset(startSupport),
+    mSupportSize(sizeSupport)
+{
+    initialize();
+}
+
+SupportBlock::SupportBlock(mem::SharedPtr<io::SeekableInputStream> inStream,
+                           const cphd::Data& data,
+                           sys::Off_T startSupport,
+                           sys::Off_T sizeSupport) :
+    mInStream(inStream),
+    mData(data),
+    mSupportOffset(startSupport),
+    mSupportSize(sizeSupport)
+{
+    initialize();
+}
+
+void SupportBlock::initialize()
+{
+    // Trusting data has the right offsets
+    mOffsets = mData.supportOffsetMap;
+    for (auto it = mOffsets.begin(); it != mOffsets.end(); ++it)
+    {
+        it->second += mSupportOffset;
+    }
+}
+
+sys::Off_T SupportBlock::getFileOffset(const std::string& id) const
+{
+    if (mOffsets.count(id) != 1)
+    {
+        throw(except::Exception(Ctxt("Invalid support array identifier number")));
+    }
+    return mOffsets.find(id)->second;
+}
+
+void SupportBlock::read(const std::string& id,
+                        size_t numThreads,
+                        const mem::BufferView<sys::ubyte>& data) const
+{
+    const size_t minSize = mData.getSupportArrayById(id).getSize();
+
+    if (data.size < minSize)
+    {
+        std::ostringstream ostr;
+        ostr << "Need at least " << minSize << " bytes but only got "
+             << data.size;
+        throw except::Exception(Ctxt(ostr.str()));
+    }
+
+    // Perform the read
+    // Compute the byte offset into this SupportArray in the CPHD file
+    // First to the start of the first support array we're going to read
+    sys::Off_T inOffset = getFileOffset(id);
+    sys::byte* dataPtr = reinterpret_cast<sys::byte*>(data.data);
+    mInStream->seek(inOffset, io::FileInputStream::START);
+    size_t size = mData.getSupportArrayById(id).getSize();
+    mInStream->read(dataPtr, size);
+
+    // Can't change endianness of compressed data right?
+    if (!sys::isBigEndianSystem() && mData.getElementSize(id) > 1)
+    {
+        cphd::byteSwap(data.data, mData.getElementSize(id),
+                       mData.getSupportArrayById(id).numRows *
+                       mData.getSupportArrayById(id).numCols,
+                       numThreads);
+    }
+}
+
+void SupportBlock::readAll(size_t numThreads,
+                           mem::ScopedArray<sys::ubyte>& data) const
+{
+    data.reset(new sys::ubyte[mSupportSize]);
+    for (auto it = mData.supportOffsetMap.begin(); it != mData.supportOffsetMap.end(); ++it)
```

Comment:
Don't you want to iterate over `supportArrayMap`? I think that's the ordered one. Actually, would it work to not loop at all and just read a chunk of `mSupportSize` bytes? Or are there padding issues or something?

Reply:
I don't think order matters here, because the FileInputStream mInStream reads from the correct file offset of the specific support array identifier and writes the data to the correct offset in the data object. So the input stream head basically jumps around in the file if the support arrays aren't in order.

The reason I didn't do a read of mSupportSize bytes is beacause each support Array can have different types of elements. So to be able to endian swap for each support array, I would need to iterate to access the specific support array's element size and number of elements.

But I'll try to think of a better solution.